charmdir.go

```diff
@@ -119,6 +120,8 @@ func ReadCharmDir(path string) (dir *CharmDir, err error) {
 			return nil, errors.New("invalid revision file")
 		}
 	}
+	version, _, err := dir.MaybeGenerateVersionString()
+	dir.version = version
```

Comment:
So this is the meat of it, in that, we always read the version string (git or file) on reading the charm directory?

Reply:
@SimonRichardson yes one of it, right. I thought this would make sense, as we initially create the `CharmDir` struct. With this, we can make sure of consistent output. 
Additionally, I added this to the struct of the charmdir, because version is a field which should belong to the struct and not only on archiving IMO.

The other change is to parse a version file if it exists.